### DISCUSSION TITLE
DataTable: Trying to update column headings

### DISCUSSION THREAD

**Discussion by leemthai** [Textualize/textual#4810] [Q&A]
Hey Guys,

I can't find any obvious way to update column headings in a DataTable.

My headings are supposed to be dynamic but the only way I can currently find to update them is clear the whole table and start again, lol, which is obviously sub-optimal. Is it possible to update the column heading labels in situ? 

loving textualize btw 💙

**Comment by TomJGooding** [comment]
There's not any obvious method for updating the column labels AFAIK.

What's your use case for updating only the column labels but not the rest of the table? Maybe this could be added as feature.

**Comment by leemthai** [comment]
Hey Tom,

I *do* actually need to update the rows as well as columns. 

The root issue I am investigating, which may be far easier to resolve, is that whenever I update the table: `self.clear(columns=True)`
and then re-add rows and columns is that I am attempting to make use of the built-in table cursor and if the cursor is in column mode at the time of the update, the column cursor jumps back to column 0 every update (when I want it to stay in place on the currently selected column)

So yesterday I was starting to investigate why the column cursor was jumping back to column 0, and I came across my code:
`self.clear(columns=True) `

and I thought aha! I am deleting columns each time and re-adding them. It *must* be that which is causing column cursor jumping.

However, I have just tried the same code but with static columns. Now I just do:
`self.clear(columns=False)`
and re-add rows. I don't re-add columns now.

And I still get the column cursor jumping back to column 0 issue. So it appears that it was not re-adding columns that is causing the issue at all.

I'm very very new to Textualize so maybe missing something very obvious here, so:
1) Do you know what I mean when I say "table cursor is jumping back to column 0"?
2) Any ideas what cause it?

There is nothing in my code to set the cursor position. The only table writing code I have is:
add_rows()
add_columns()

Cheers,

**Reply by leemthai** [reply]
Just to confirm something. It *is* the call of self.clear() that causes the column cursor to jump back to zero. If I leave self.clear() out of my code, and let new rows be added to the bottom of the table, the cursor stays in place.

Also confirmed, if the cursor is in 'row' mode instead, the same thing happens with self.clear() is called. The cursor goes back to row zero in this case.

So it appears that datatable.clear() always resets the table cursor back to zero?

**Reply by leemthai** [reply]
Which is perfectly sensible of course. An empy table's cursor should be at (0,0).

**Comment by leemthai** [comment]
This is my current solution:

```
        old_column = self.cursor_column
        self.clear(columns=True)  # Clear the table rows and columns coz need to regenerate both every time
        self.generate_columns()
        self.move_cursor(column=old_column) # Note must do this AFTER generate_columns
```

So I just preserve the selected column number, clear the table, regenerate the columns (in my app this will always be the same number of columns, that never changes), and reinstate the column. The row generation comes a bit later in the code.

This *appears* to work well and look smooth. Please let me know if this approach is likely to break anything :)

Thanks,

**Reply by TomJGooding** [reply]
> So it appears that datatable.clear() always resets the table cursor back to zero? ... Which is perfectly sensible of course. An empy table's cursor should be at (0,0).

Yes exactly. When you remove a row from the table, the cursor coordinate is validated accordingly. For example, let's say rather than clearing the whole table, you were removing only the last row:

```
| Col 0 |
|-------|
| Row 0 |
| Row 1 | 
| Row 2 |  <- Cursor at Coordinate(2, 0)

After removing the last row in the table:

| Col 0 |
|-------|
| Row 0 |
| Row 1 |  <- Cursor now at Coordinate(1, 0)
```

> This is my current solution [...] So I just preserve the selected column number, clear the table, regenerate the columns [...[ and reinstate the column

It might be better to save the current `cursor_coordinate` and restore that after adding your rows?

```python
from textual.app import App, ComposeResult
from textual.widgets import DataTable, Footer


class ExampleApp(App):
    BINDINGS = [("n", "new_table", "Create new table")]

    table_number = 0

    def compose(self) -> ComposeResult:
        yield DataTable()
        yield Footer()

    def on_mount(self) -> None:
        self.populate_table()

    def populate_table(self) -> None:
        table = self.query_one(DataTable)

        # Save the current table cursor coordinate
        cursor_coordinate = table.cursor_coordinate

        table.clear(columns=True)

        self.table_number += 1
        table.add_columns(*(f"Table #{self.table_number}" for _ in range(10)))
        for _ in range(10):
            table.add_row(*(f"Table #{self.table_number}" for _ in range(10)))

        # Restore the cursor coordinate
        table.cursor_coordinate = cursor_coordinate

    def action_new_table(self) -> None:
        self.populate_table()


if __name__ == "__main__":
    app = ExampleApp()
    app.run()
```

**Reply by leemthai** [reply]
Right... that looks good. I guess that was my first thought but then I realized the new table could have less rows than the old one.... is it ok to restore to a position beyond the last row?

**Reply by leemthai** [reply]
Ah yes I just tested it. If the new table has less rows, the row cursor will go back to 0. Very good :) I will close this now. Thanks for the help!

**Reply by TomJGooding** [reply]
Glad this helped! Yes the `cursor_coordinate` is a [reactive with validation](https://textual.textualize.io/guide/reactivity/#validation) which clamps the coordinate, so you don't need to worry about assigning an out of range row/column.
